// @greatmap/vue-webgis/src/gis/api/modules/buffer-analysis.js
import {
  createOverlayAnalysis as baseCreateOverlayAnalysis,
  createSummaryAnalysis as baseCreateSummaryAnalysis
} from '../analysis-service';
import { SERVER_CONFIG } from '../../config/server';

const arcgisOrigin = SERVER_CONFIG.arcgisOrigin;
const superDataPath = SERVER_CONFIG.superDataPath;

/**
 * 创建叠加分析
 * @param {String} serviceUrl 服务 URL，创建服务返回
 * @param {String} serviceId 服务 ID，创建服务返回
 * @param {String} name overlay 服务名称
 * @param {String} bufferName buffer 服务名称
 * @param {String} serviceCode 服务代码，例如 X620100DC2019GXZQ
 * @returns {Promise}
 */
export function createOverlayAnalysis({
  serviceUrl,
  serviceId,
  name,
  bufferName,
  serviceCode
}) {
  return baseCreateOverlayAnalysis({
    serviceUrl,
    serviceId,
    name,
    inputLayer: {
      url: `${arcgisOrigin}/server/rest/services/DataStoreCatalogs/${superDataPath}/BigDataCatalogServer/${serviceCode}`,
      name: `${superDataPath}-${serviceCode}`
    },
    overlayLayer: {
      url: `${arcgisOrigin}/server/rest/services/Hosted/${bufferName}/FeatureServer/0`,
      name: bufferName
    }
  });
}

/**
 * 创建汇总分析
 * @param {String} serviceUrl 服务 URL，创建服务返回
 * @param {String} serviceId 服务 ID，创建服务返回
 * @param {String} name 汇总分析名称
 * @param {String} overlayName 叠加分析名称
 * @return {Promise}
 */
export function createSummaryAnalysis({
  serviceUrl,
  serviceId,
  name,
  overlayName
}) {
  return baseCreateSummaryAnalysis({
    serviceUrl,
    serviceId,
    name,
    inputLayer: {
      url: `${arcgisOrigin}/server/rest/services/Hosted/${overlayName}/FeatureServer/0`
    }
  });
}
