// @greatmap/vue-webgis/src/gis/api/service.js
import request from '../utils/request';
import { SERVER_CONFIG } from '../config/server';

/**
 * 获取服务列表
 * @returns {Promise}
 */
export function getServiceList(type = 'Map Service') {
  return request({
    url: SERVER_CONFIG.arcgisServiceList,
    method: 'post',
    data: {
      q: `owner:${
        SERVER_CONFIG.arcgisUsername
      } orgid:0123456789ABCDEF (type:("${type}")
      -typekeywords:("Tiled" OR "Hosted") -type:("Web Map" OR "Web Mapping Application" OR "Shapefile"))
      ownerfolder:root -type:("Code Attachment" OR "Featured Items")
      -typekeywords:("MapAreaPackage") -type:("Map Area")`,
      f: 'json',
      num: 999
    },
    analysisToken: true
  });
}

/**
 * 获取服务配置
 * @returns {Promise}
 */
export function getServiceConfig(code, type = 'MapServer', id = '') {
  return request({
    baseURL: SERVER_CONFIG.arcgisQuery,
    url: `/arcgis/rest/services/${code}/${type}/${id}`,
    method: 'post',
    data: {
      f: 'json'
    }
  });
}

/**
 * 获取服务详情（features，attributes）
 *
 * @param {string} serviceCode 服务地址，例如：X510183TDXZG2017XZQ
 * @param {string} layerId 图层的 id，例如：0
 * @param {string} epsgNum 参考系，例如：4499
 * @param {string} geometry 几何范围，JSON 字符串
 * @param {string} geometryType geometry 类型，例如：esriGeometryPolygon|esriGeometryEnvelope
 * @param {string} [where] 查询语句，例如：BSM='123'
 * @param {boolean} [returnGeometry=true] 是否返回 geometry 数据
 * @param {string} [outFields=*] 输出 fields，例如：'BSM,DLTB'
 * @param {String} [groupByFieldsForStatistics] 字段组合分析
 * @param {String} [outStatistics] 返回分析的字段
 * @return {Promise}
 */
export function getServiceDetail({
  serviceCode,
  layerId,
  epsgNum,
  geometry,
  geometryType,
  where = '1=1',
  outFields = '*',
  returnGeometry = true,
  groupByFieldsForStatistics = '',
  outStatistics
}) {
  return request({
    baseURL: SERVER_CONFIG.arcgisQuery,
    url: `/arcgis/rest/services/${serviceCode}/MapServer/${layerId}/query`,
    method: 'post',
    data: {
      f: 'json',
      returnGeometry,
      spatialRel: 'esriSpatialRelIntersects',
      geometry,
      geometryType,
      outFields,
      inSR: epsgNum,
      outSR: epsgNum,
      groupByFieldsForStatistics,
      outStatistics,
      where
    }
  });
}
