// @greatmap/vue-webgis/src/gis/api/token.js
import { analysisAuth, gisServerAuth } from '../utils/gis-server-auth';
import { SERVER_CONFIG } from '../config/server';
import request from '../utils/request';

/**
 * 获取查询用的 token
 * @description
 * process 作用为防重复请求。
 * @returns {Promise}
 */
export function getToken() {
  return (getToken.process = getToken.process
    ? getToken.process
    : createToken().then((res) => {
        getToken.process = null;
        if (res && res.token) {
          gisServerAuth.setToken(res);
        }
      }));
}

/**
 * 获取分析用的 token
 * @description
 * process 作用为防重复请求。
 * @returns {Promise}
 */
export function getAnalysisToken() {
  return (createToken.process = createToken.process
    ? createToken.process
    : createToken(SERVER_CONFIG.host).then((res) => {
        createToken.process = null;
        if (res && res.token) {
          analysisAuth.setToken(res);
        }
      }));
}

/**
 * 创建 token
 * @params {String} referer 生成 token 的 referer 参数
 * 1. referer: serverUrl 生成查询 token
 * 2. referer: host 生成分析 token
 */
function createToken(referer = SERVER_CONFIG.host) {
  return request({
    url: SERVER_CONFIG.arcgisToken,
    method: 'post',
    data: {
      username: SERVER_CONFIG.arcgisUsername,
      password: SERVER_CONFIG.arcgisPassword,
      referer,
      client: 'referer',
      f: 'json'
    }
  });
}
