// @greatmap/vue-webgis/lib/chunk-bc626433.js
/*!
 * vue-webgis v0.7.3
 */

import 'core-js/modules/es6.function.name';
import { _ as _createClass, a as _classCallCheck, b as _defineProperty } from './chunk-82f2c815.js';

function getLocalData(name) {
  var data = localStorage.getItem(name);

  if (data) {
    return JSON.parse(data);
  } else {
    return data;
  }
}
function setLocalData(name, data) {
  localStorage.setItem(name, JSON.stringify(data));
}

var Auth =
/*#__PURE__*/
function () {
  function Auth(tokenName) {
    _classCallCheck(this, Auth);

    _defineProperty(this, "name", '');

    this.name = tokenName;
  }

  _createClass(Auth, [{
    key: "getToken",
    value: function getToken() {
      var data = getLocalData(this.name);

      if (data && data.expires > Date.now()) {
        return data.token;
      } else {
        return null;
      }
    }
  }, {
    key: "setToken",
    value: function setToken(token) {
      setLocalData(this.name, token);
    }
  }]);

  return Auth;
}();

var gisServerAuth = new Auth('GIS_SERVER_TOKEN');
var analysisAuth = new Auth('ANALYSIS_TOKEN');

var SERVER_CONFIG = {
  host: location.host,
  arcgis: 'http://wh.greatmap.com.cn:16442',
  arcgisQuery: 'http://wh.greatmap.com.cn:16443',
  arcgisOrigin: 'https://gis120.arcgisonline.cn',
  superDataPath: 'bigDataFileShares_XZSJ',
  arcgisService: '',
  arcgisRestService: '',
  arcgisToken: '',
  arcgisServiceList: '',
  arcgisUsername: 'admin',
  arcgisPassword: 'arcgis123'
};
function updateServerAPI() {
  var config = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  Object.assign(SERVER_CONFIG, config);
  Object.assign(SERVER_CONFIG, {
    arcgisService: SERVER_CONFIG.arcgisQuery + '/arcgis/services/',
    arcgisRestService: SERVER_CONFIG.arcgisQuery + '/arcgis/rest/services/',
    arcgisToken: SERVER_CONFIG.arcgis + '/portal/sharing/rest/generateToken',
    arcgisServiceList: SERVER_CONFIG.arcgis + '/portal/sharing/rest/search'
  }, config);
}
updateServerAPI();

export { SERVER_CONFIG as S, analysisAuth as a, gisServerAuth as g, updateServerAPI as u };
