// @greatmap/vue-webgis/lib/chunk-dd3bd2eb.js
/*!
 * vue-webgis v0.7.3
 */

import { _ as _createClass, a as _classCallCheck, b as _defineProperty } from './chunk-82f2c815.js';

var ToggleActive =
/*#__PURE__*/
function () {
  function ToggleActive() {
    _classCallCheck(this, ToggleActive);

    _defineProperty(this, "isActivated", false);
  }

  _createClass(ToggleActive, [{
    key: "toggleActive",
    value: function toggleActive() {
      var state = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : !this.isActivated;

      if (state) {
        for (var _len = arguments.length, params = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
          params[_key - 1] = arguments[_key];
        }

        this.activate.apply(this, params);
      } else {
        this.deactivate();
      }
    }
  }, {
    key: "activateHandle",
    value: function activateHandle() {}
  }, {
    key: "deactivateHandle",
    value: function deactivateHandle() {}
  }, {
    key: "activate",
    value: function activate() {
      this.isActivated && this.deactivate();
      this.isActivated = true;
      this.activateHandle.apply(this, arguments);
    }
  }, {
    key: "deactivate",
    value: function deactivate() {
      if (!this.isActivated) return;
      this.isActivated = false;
      this.deactivateHandle();
    }
  }]);

  return ToggleActive;
}();

export { ToggleActive as T };
