// @greatmap/vue-webgis/lib/diff-analysis-e4426044.js
/*!
 * vue-webgis v0.7.3
 */

import 'vue';
import 'core-js/modules/es6.function.name';
import 'core-js/modules/es7.array.includes';
import 'core-js/modules/es6.string.includes';
import 'core-js/modules/es6.array.find';
import 'core-js/modules/web.dom.iterable';
import 'core-js/modules/es6.object.keys';
import './chunk-8430f437.js';
import 'core-js/library/fn/array/is-array';
import './chunk-2f41c7fa.js';
import 'core-js/library/fn/array/from';
import 'core-js/library/fn/is-iterable';
import 'core-js/library/fn/object/define-property';
import { a as _classCallCheck, _ as _createClass } from './chunk-82f2c815.js';
import 'core-js/library/fn/symbol/iterator';
import { _ as _inherits, a as _possibleConstructorReturn, b as _getPrototypeOf } from './chunk-537b1685.js';
import 'core-js/library/fn/symbol';
import 'core-js/library/fn/object/get-prototype-of';
import 'core-js/library/fn/object/set-prototype-of';
import 'core-js/library/fn/object/create';
import 'core-js/modules/es6.regexp.split';
import { S as SERVER_CONFIG } from './chunk-bc626433.js';
import 'axios';
import 'qs/lib/stringify';
import './chunk-81d749a9.js';
import { c as createOverlayAnalysis$1, b as createAnalysisService, g as getAnalysisResult, B as BaseAnalysis } from './chunk-fc6359dc.js';

var arcgisOrigin = SERVER_CONFIG.arcgisOrigin;
var superDataPath = SERVER_CONFIG.superDataPath;
function createOverlayAnalysis(_ref) {
  var serviceUrl = _ref.serviceUrl,
      serviceId = _ref.serviceId,
      name = _ref.name,
      inputServiceCode = _ref.inputServiceCode,
      inputFilter = _ref.inputFilter,
      overlayServiceCode = _ref.overlayServiceCode,
      overlayFilter = _ref.overlayFilter,
      _ref$overlayType = _ref.overlayType,
      overlayType = _ref$overlayType === void 0 ? 'Erase' : _ref$overlayType;
  var inputLayer = {
    url: "".concat(arcgisOrigin, "/server/rest/services/DataStoreCatalogs/").concat(superDataPath, "/BigDataCatalogServer/").concat(inputServiceCode),
    name: "".concat(superDataPath, "-").concat(inputServiceCode)
  };

  if (inputFilter) {
    inputLayer.filter = inputFilter;
  }

  var overlayLayer = {
    url: "".concat(arcgisOrigin, "/server/rest/services/DataStoreCatalogs/").concat(superDataPath, "/BigDataCatalogServer/").concat(overlayServiceCode),
    name: "".concat(superDataPath, "-").concat(overlayServiceCode)
  };

  if (overlayFilter) {
    overlayLayer.filter = overlayFilter;
  }

  return createOverlayAnalysis$1({
    serviceUrl: serviceUrl,
    serviceId: serviceId,
    name: name,
    overlayType: overlayType,
    inputLayer: inputLayer,
    overlayLayer: overlayLayer
  });
}

var DiffAnalysis =
/*#__PURE__*/
function (_BaseAnalysis) {
  _inherits(DiffAnalysis, _BaseAnalysis);

  function DiffAnalysis(map) {
    _classCallCheck(this, DiffAnalysis);

    return _possibleConstructorReturn(this, _getPrototypeOf(DiffAnalysis).call(this, map));
  }

  _createClass(DiffAnalysis, [{
    key: "doAnalysis",
    value: function doAnalysis(_ref) {
      var _this = this;

      var inputServiceCode = _ref.inputServiceCode,
          inputFilter = _ref.inputFilter,
          overlayServiceCode = _ref.overlayServiceCode,
          overlayFilter = _ref.overlayFilter;
      var serviceName = 'diffAnalysis_overlay_' + Date.now();
      return createAnalysisService(serviceName).then(function (res) {
        return createOverlayAnalysis({
          serviceUrl: res.serviceurl,
          serviceId: res.itemId,
          name: serviceName,
          inputServiceCode: inputServiceCode,
          inputFilter: inputFilter,
          overlayServiceCode: overlayServiceCode,
          overlayFilter: overlayFilter
        }).then(function (res) {
          return _this.getProgress('OverlayLayers', res.jobId).then(function () {
            return getAnalysisResult(serviceName, _this._wkid).then(function (res) {
              _this._map.interactions.zoom.toFeatures(_this._map.renderFeatureData(res));

              return res;
            });
          });
        });
      });
    }
  }]);

  return DiffAnalysis;
}(BaseAnalysis);

export default DiffAnalysis;
