// @greatmap/vue-webgis/lib/image-layer-ad3ffee6.js
/*!
 * vue-webgis v0.7.3
 */

import 'core-js/modules/es6.function.name';
import 'core-js/library/fn/object/define-property';
import { _ as _createClass, a as _classCallCheck, b as _defineProperty } from './chunk-82f2c815.js';
import 'ol/source/TileWMS';
import 'ol/layer/Tile';
import 'ol/source/XYZ';
import 'ol/tilegrid/TileGrid';
import 'ol/source/ImageArcGISRest';
import 'ol/layer/Image';
import './chunk-bc626433.js';
import { L as LayerManagement } from './chunk-2b89b3e5.js';

var ImageLayer =
/*#__PURE__*/
function () {
  function ImageLayer(map) {
    _classCallCheck(this, ImageLayer);

    _defineProperty(this, "_imageLayers", void 0);

    _defineProperty(this, "_olMap", void 0);

    this._olMap = map.olMap;
  }

  _createClass(ImageLayer, [{
    key: "change",
    value: function change(serviceCode, serverType, serverId) {
      this._imageLayers && this.remove();
      this._imageLayers = LayerManagement.createImageLayer({
        serviceCode: serviceCode,
        serverType: serverType,
        serverId: serverId
      });

      this._olMap.addLayer(this._imageLayers);
    }
  }, {
    key: "remove",
    value: function remove() {
      this._olMap.removeLayer(this._imageLayers);

      this._imageLayers = null;
    }
  }]);

  return ImageLayer;
}();

export default ImageLayer;
