// @greatmap/vue-webgis/src/mixins/private-event-hub.js
/**
 * 私有的事件中心注册，仅用于单个的 webgis 组件上（在分屏时应用于单个内）
 * @description
 * 1. **用于 webgis 内部的事件传递**
 * 2. **`$_eventHub` 用于处理 provide 与 inject 同时使用的情况**
 * 3. **`privateEventHubProvide` 在一个范围内仅使用一次**
 *
 * @example
 * ```
 * import {privateEventHubProvide, privateEventHubInject} from './private-event-hub';
 *
 * export default {
 *   mixins: [privateEventHubProvide, privateEventHubInject], // 加入
 *   methods: {
 *     doSomeThing() {
 *       this.$eventHub.$emit('someEvent', 'test'); // 使用
 *     }
 *   }
 * }
 * ```
 */
import Vue from 'vue';
import { warnMessage } from '../utils/error-handler';

export const privateEventHubProvide = {
  provide() {
    return {
      $_eventHub: this.$eventHub
    };
  },
  beforeCreate() {
    // 禁用重复注册
    if (this.$eventHub) {
      return warnMessage('私有的事件中心在一个范围内仅能存在一个!');
    }
    this.$eventHub = new Vue();
  }
};

export const privateEventHubInject = {
  inject: {
    $_eventHub: {
      type: Object,
      default: null
    }
  },
  created() {
    this.$eventHub = this.$_eventHub || this.$data.$evenHub;
  }
};
