// @greatmap/vue-webgis/src/mixins/webgis-init.js
/**
 * 此混入对象作为当前的组件库内使用
 * @description
 * **此处涉及分屏的状态管理，逻辑复杂**
 * 1. webgisProvide 在每一个 webgis 组件中，仅存在一个，所以在分屏中会有很多的实例
 * 2. webgisInject 在每一个 webgis 组件中，仅作为 webgisProvide 的子组件使用，非子组件不能使用
 */
import webgisInitCommon from './webgis-init-common';
import {
  privateEventHubInject,
  privateEventHubProvide
} from './private-event-hub';

const eventName = 'privateWebgisReady';

export const webgisProvide = {
  mixins: [webgisInitCommon, privateEventHubProvide],
  methods: {
    initWebgis(webgis) {
      this.$emit('init', webgis); // emit parent
      this.$eventHub.$emit(eventName, webgis); // emit inject
      this.$eventHub.$webgis = webgis; // hook
      this._afterInitWebgis(webgis);
    }
  }
};

export const webgisInject = {
  mixins: [webgisInitCommon, privateEventHubInject],
  methods: {
    _initWebgis(webgis) {
      // cancel event listener
      this.$eventHub.$off(eventName, this._initWebgis);
      this._afterInitWebgis(webgis);
    }
  },
  created() {
    // if `$webgis` ready, don't need listener
    const webgis = this.$eventHub.$webgis;
    if (webgis) {
      this.$webgis = webgis;
      this._initWebgis(webgis);
    } else {
      this.$eventHub.$on(eventName, this._initWebgis);
    }
  }
};
