// @greatmap/vue-webgis/src/gis/core/interactions/drag-zoom.js
import OlDragZoom from 'ol/interaction/DragZoom';
import { mouseOnly } from 'ol/events/condition';
import ToggleActive from '../common/toggle-active';
import classMixin from '../../utils/class-mixin';
import GISEvent from '../common/gis-event';

/**
 * 拖动缩放
 */
export default class DragZoom extends classMixin(GISEvent, ToggleActive) {
  _map;
  _olMap;
  _options = {
    out: null,
    active: false,
    condition: mouseOnly
  };

  constructor(map) {
    super('boxstart,boxend');

    this._map = map;
    this._olMap = map.olMap;
  }

  /**
   * 切换状态
   */
  toggleActive(state, type) {
    const out = type !== 'zoom-in';
    if (state) {
      this.activate(out);
    } else {
      this.deactivate();
    }
  }

  // 激活 extends ToggleActive
  activateHandle(out) {
    this._options.out = out;
    this._olMap.addInteraction(
      (this._dragZoom = new OlDragZoom(this._options))
    );
    this._dragZoom.on('boxstart', (e) => this.dispatch('boxstart', e));
    this._dragZoom.on('boxend', (e) => this.dispatch('boxend', e));
    this._map.toggleCursorStyle(out ? 'zoom-out' : 'zoom-in');
  }

  // 关闭 extends ToggleActive
  deactivateHandle() {
    this._options.out = null;
    this._olMap.removeInteraction(this._dragZoom);
    this._map.toggleCursorStyle();
  }
}
