// @greatmap/vue-webgis/src/gis/config/server.js
/**
 * 服务地址配置
 */

/**
 * @typedef {object} ServerAPI
 * @property {string} [host] 域名地址，默认取 location.host
 * @property {string} [arcgis] arcgis 接口地址，例如 http://wh.greatmap.com.cn:16442
 * @property {string} [arcgisQuery] arcgis 查询接口地址，例如 http://wh.greatmap.com.cn:16443
 * @property {string} [arcgisOrigin] arcgis 原始域名，例如 https://gis120.arcgisonline.cn
 * @property {string} [arcgisService] arcgis/service 接口
 * @property {string} [arcgisRestService] arcgis/rest/service 接口
 * @property {string} [superDataPath] arcgis 大数据目录
 * @property {string} [arcgisToken] arcgis 获取 token
 * @property {string} [arcgisServiceList] arcgis 获取服务列表
 * @property {string} [arcgisUsername] arcgis 用户名
 * @property {string} [arcgisPassword] arcgis 密码
 */

/**@type ServerAPI 服务地址配置 */
export const SERVER_CONFIG = {
  host: location.host,
  arcgis: 'http://wh.greatmap.com.cn:16442',
  arcgisQuery: 'http://wh.greatmap.com.cn:16443',
  arcgisOrigin: 'https://gis120.arcgisonline.cn',
  superDataPath: 'bigDataFileShares_XZSJ',
  arcgisService: '',
  arcgisRestService: '',
  arcgisToken: '',
  arcgisServiceList: '',
  arcgisUsername: 'admin',
  arcgisPassword: 'arcgis123'
};

/**
 * 更新服务地址
 * @param {ServerAPI} config
 */
export function updateServerAPI(config = {}) {
  Object.assign(SERVER_CONFIG, config);
  Object.assign(
    SERVER_CONFIG,
    {
      arcgisService: SERVER_CONFIG.arcgisQuery + '/arcgis/services/',
      arcgisRestService: SERVER_CONFIG.arcgisQuery + '/arcgis/rest/services/',
      arcgisToken: SERVER_CONFIG.arcgis + '/portal/sharing/rest/generateToken',
      arcgisServiceList: SERVER_CONFIG.arcgis + '/portal/sharing/rest/search'
    },
    config
  );
}

// 初始化一次
updateServerAPI();
