// @greatmap/vue-webgis/src/common.js
import Vue from 'vue';

// 主组件
import WebGIS from './components';

// 其他组件
import WebGISDiffAnalysis from '../packages/diff-analysis';

// styles
import './styles/public.scss';

// mixins
import webgisInitCommon from './mixins/webgis-init-common';

// 对外事件中心 - 全局的（存在于应用中共享）
const eventHub = new Vue();
const eventName = 'publicWebgisReady';

// webgis 初始化注册用的 mixin 对象
export const webgisProvidePublic = {
  mixins: [webgisInitCommon],
  methods: {
    initWebgis(webgis) {
      eventHub.$emit(eventName, webgis); // emit inject
      eventHub.$webgis = webgis; // hook
      this._afterInitWebgis(webgis);
    }
  }
};

// webgis 初始化注入用的 mixin 对象
export const webgisInjectPublic = {
  mixins: [webgisInitCommon],
  methods: {
    _initWebgis(webgis) {
      // cancel event listener
      eventHub.$off(eventName, this._initWebgis);
      this._afterInitWebgis(webgis);
    }
  },
  created() {
    // if `$webgis` ready, don't need listener
    const webgis = eventHub.$webgis;
    if (webgis) {
      this.$webgis = webgis;
      this._initWebgis(webgis);
    } else {
      eventHub.$on(eventName, this._initWebgis);
    }
  }
};

// 对外事件中心 - 私有的（存在与 webgis 当前的组件范围内）
export {
  webgisProvide as webgisProvidePrivate,
  webgisInject as webgisInjectPrivate
} from './mixins/webgis-init';

// 更新服务地址配置接口
export { updateServerAPI } from './gis/config/server';

// Vue.use()
WebGIS.install = (Vue) => {
  Vue.component('WebGIS', WebGIS);
  Vue.use(WebGISDiffAnalysis);
};

export { WebGISDiffAnalysis };

// Vue.component()
export default WebGIS;
